v4 (#1)
Skn0tt/adapters: packages/netlify/README.md

   ```diff lang="js"
     // astro.config.mjs
     import { defineConfig } from 'astro/config';
   + import netlify from '@astrojs/netlify';

     export default defineConfig({
   +   output: 'server',
   +   adapter: netlify(),
     });
   ```

#### Accessing edge context from your site

Netlify Edge Functions provide a [context object](https://docs.netlify.com/edge-functions/api/#netlify-specific-context-object) including metadata about the request, such as a user’s IP, geolocation data, and cookies.

This can be accessed via the `Astro.locals.netlify.context` object:

```astro
---

> const { geo: { city } } = Astro.locals.netlify.context

Reviewer: Why not `const { city } = Astro.locals.netlify.context.geo`?

Author: no strong preference between the two. Side-note: Do you think `Astro.locals.netlify.context` is a bit verbose? Between these, do you have any preference?
- `Astro.locals.netlifyContext`
- `Astro.locals.context`
- `Astro.locals.netlify`

Reviewer: The reason I suggested changing it is because the syntax is more confusing. For the other question, what el;se is `Astro.locals` used for? If `context` is the only other thing we're adding, I'd say just make it `context` at the top level

Author: It's what Astro uses to make transport data from the middleware into the SSR rendering: https://docs.astro.build/en/reference/api-reference/#astrolocals
So all other properties on there will be user-configured. `Astro.locals.context` is probably the best, yeah

Author: one reason that `context` might not be ideal is that Astro already has a `context` parameter they're passing into middleware. Accessing Netlify's context would look like this then:
```ts
import { defineMiddleware } from "astro:middleware";

export const onRequest = defineMiddleware(async (context, next) => {
  context.locals.populatedFromMiddleware = "I was populated from a middleware!";

  context.locals.context // <-- might be confusing
  const resp = (await next()) as Response;
  resp.headers.set("x-astro-middleware", "true");
  return resp;
});
```

this feels a little confusing? makes me think we should keep `netlify` in the name in some way

   ```diff lang="js"
     // astro.config.mjs
     import { defineConfig } from 'astro/config';
   + import netlify from '@astrojs/netlify';

     export default defineConfig({
   +   output: 'server',
   +   adapter: netlify(),
     });
   ```

### Accessing edge context from your site

> Netlify Edge Functions provide a [context object](https://docs.netlify.com/edge-functions/api/#netlify-specific-context-object) including metadata about the request, such as a user’s IP, geolocation data, and cookies.

Reviewer 2: [sand] For clarity

```suggestion
Netlify Edge Functions provide a [context object](https://docs.netlify.com/edge-functions/api/#netlify-specific-context-object) that includes metadata about the request such as a user’s IP, geolocation data, and cookies.
```

> This can be accessed via the `Astro.locals.netlify.context` object:

Reviewer 2: [pebble] Although this is a readme, applying our docs style guide rule here to [avoid the word `via`](https://netlify-handbook.netlify.app/product/documentation/docs-style-guide/terminology/#via)

```suggestion
This can be accessed through the `Astro.locals.netlify.context` object:
```

```astro
---
const { geo: { city } } = Astro.locals.netlify.context
---
<h1>Hello there, friendly visitor from {city}!</h1>
```

If you're using TypeScript, you can get proper typings by updating your `src/env.d.ts` to use `NetlifyLocals`:

```ts
// src/env.d.ts
/// <reference path="../.astro/types.d.ts" />
/// <reference types="astro/client" />

type NetlifyLocals = import('@astrojs/netlify').NetlifyLocals

declare namespace App {
  interface Locals extends NetlifyLocals {
    ...
  }
}
```

### Image CDN

This adapter integrates your site with [Netlify Image CDN](https://docs.netlify.com/image-cdn/), transforming images on-the-fly without impacting build times.
It's implemented using an [Astro Image Service](https://docs.astro.build/en/reference/image-service-reference/) under the hood.

> Note that this adapter does not support `image.domains` and `image.remotePatterns` config properties in your Astro config - use the `remote_images` field in `netlify.toml` instead: [Netlify Image CDN - Remote Path](https://docs.netlify.com/image-cdn/overview/#remote-path).

Reviewer 2: [dust] Could put this note on its own line (create a new paragraph) to make sure it's easy to see? Right now when it's all one paragraph it blends together a bit. You could also use the `**Note**` formatting.

```suggestion
This adapter integrates your site with [Netlify Image CDN](https://docs.netlify.com/image-cdn/), transforming images on-the-fly without impacting build times.
It's implemented using an [Astro Image Service](https://docs.astro.build/en/reference/image-service-reference/) under the hood.

> **Note**
> This adapter does not support `image.domains` and `image.remotePatterns` config properties in your Astro config - use the `remote_images` field in `netlify.toml` instead: [Netlify Image CDN - Remote Path](https://docs.netlify.com/image-cdn/overview/#remote-path).
```

Author: good idea! looks a bit like the "warning" callouts we have in our docs :D

   ```diff lang="js"
     // astro.config.mjs
     import { defineConfig } from 'astro/config';
   + import netlify from '@astrojs/netlify';

     export default defineConfig({
   +   output: 'server',
   +   adapter: netlify(),
     });
   ```

## Usage

[Read the full deployment guide here.](https://docs.astro.build/en/guides/deploy/netlify/)

After [performing a build](https://docs.astro.build/en/guides/deploy/#building-your-site-locally) the `.netlify/` folder will contain [Netlify Functions](https://docs.netlify.com/functions/overview/) in the `.netlify/functions-internal/` and [Netlify Edge Functions](https://docs.netlify.com/edge-functions/overview/) in the `.netlify/edge-functions/` folder.

Now you can deploy. Install the [Netlify CLI](https://docs.netlify.com/cli/get-started/) and run:

```sh
netlify deploy
```

> The [Netlify Blog post on Astro](https://www.netlify.com/blog/how-to-deploy-astro/) and the [Netlify Documentation](https://docs.netlify.com/integrations/frameworks/astro/) provide more information on how to use this integration to deploy to Netlify.

Reviewer 2: [sand] "Documentation" sounded a bit weird, think it flows better to just say "Docs", especially since that's how we refer to the site

```suggestion
The [Netlify Blog post on Astro](https://www.netlify.com/blog/how-to-deploy-astro/) and the [Netlify Docs](https://docs.netlify.com/integrations/frameworks/astro/) provide more information on how to use this integration to deploy to Netlify.
```

> **Note**
> You can still include a `public/_redirects` file for manual redirects. Any redirects you specify in the redirects config are appended to the end of your own.

### Caching SSR Pages using Fine-Grained Cache Control

With [fine-grained cache control](https://www.netlify.com/blog/swr-and-fine-grained-cache-control/), Netlify supports
standard caching headers like `CDN-Cache-Control` or `Vary`.

> This allows you to implement e.g. Time to live (TTL) or [Stale while revalidate (SWR)](https://docs.netlify.com/platform/caching/#stale-while-revalidate-directive) caching.

Reviewer: ```suggestion
This allows you to implement e.g. time to live (TTL) or [stale while revalidate (SWR)](https://docs.netlify.com/platform/caching/#stale-while-revalidate-directive) caching.
```

> **Note**
> You can still include a `public/_redirects` file for manual redirects. Any redirects you specify in the redirects config are appended to the end of your own.

### Caching SSR Pages using Fine-Grained Cache Control

With [fine-grained cache control](https://www.netlify.com/blog/swr-and-fine-grained-cache-control/), Netlify supports
standard caching headers like `CDN-Cache-Control` or `Vary`.

> Use these to add caching to your SSR Pages:

Reviewer: Maybe mention stale-whike-revalidate specifically?

Author: done in https://github.com/Skn0tt/adapters/pull/1/commits/3bda429b71b0b379dbbad0593bfa4d3d74d0b39b, let me know how you like that

```astro
---
// src/pages/index.astro
import Layout from '../components/Layout.astro';

Astro.response.headers.set('CDN-Cache-Control', "public, max-age=45, must-revalidate")
---

<Layout title="Astro on Netlify">
  {new Date(Date.now())}
</Layout>
```

## Usage

[Read the full deployment guide here.](https://docs.astro.build/en/guides/deploy/netlify/)

After [performing a build](https://docs.astro.build/en/guides/deploy/#building-your-site-locally) the `.netlify/` folder will contain [Netlify Functions](https://docs.netlify.com/functions/overview/) in the `.netlify/functions-internal/` and [Netlify Edge Functions](https://docs.netlify.com/edge-functions/overview/) in the `.netlify/edge-functions/` folder.

Now you can deploy. Install the [Netlify CLI](https://docs.netlify.com/cli/get-started/) and run:

> 

Reviewer: This is confusing, because it tells you to perform a build and then says you can deploy, but the dploy command includes `--build`

Author: done in 95fad12da6ad0cb73ab3eaed7da7f857dcdeefcc

> **Note**
> You can still include a `public/_redirects` file for manual redirects. Any redirects you specify in the redirects config are appended to the end of your own.

### Caching SSR Pages using Fine-Grained Cache Control

With [fine-grained cache control](https://www.netlify.com/blog/swr-and-fine-grained-cache-control/), Netlify supports

> standard caching headers like `CDN-Cache-Control` or `Vary`.

Reviewer 2: [dust] Spacing looks weird here but I think it's probably just how this file displays? Mentioning it in case it needs to be fixed but I think it should be fine?

Author: I think it's fine, yeah

The [Netlify Blog post on Astro](https://www.netlify.com/blog/how-to-deploy-astro/) and the [Netlify Documentation](https://docs.netlify.com/integrations/frameworks/astro/) provide more information on how to use this integration to deploy to Netlify.

> TODO: update astro template after this is released

Reviewer: Probably don't want this in the PR!
